server/src/main/java/org/elasticsearch/action/admin/indices/mapping/get/GetFieldMappingsResponse.java

```diff
@@ -20,27 +20,77 @@
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentFragment;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.common.xcontent.XContentType;
 import org.elasticsearch.index.mapper.Mapper;
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.unmodifiableMap;
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
 
 /** Response object for {@link GetFieldMappingsRequest} API */
 public class GetFieldMappingsResponse extends ActionResponse implements ToXContentFragment {
 
+    private static final ParseField MAPPINGS = new ParseField("mappings");
+
+    private static final ObjectParser<Builder, String> PARSER =
+        new ObjectParser<>(MAPPINGS.getPreferredName(), true, Builder::new);
+
+    static {
+        PARSER.declareField((p, b, index) -> {
+            final Map<String, Object> map = p.map();
+            final Map<String, Map<String, FieldMappingMetaData>> typeMappings = new HashMap<>();
+            for (Map.Entry<String, Object> typeObjectEntry : map.entrySet()) {
+                final Map<String, FieldMappingMetaData> fieldsMapping = new HashMap<>();
+                typeMappings.put(typeObjectEntry.getKey(), fieldsMapping);
+
+                final Object o1 = typeObjectEntry.getValue();
+                if (!(o1 instanceof Map)) {
+                    throw new ParsingException(p.getTokenLocation(),
+                        "Nested type mapping at " + index + "." + typeObjectEntry.getKey() + " is not found");
+                }
+                final Map<String, Object> map2 = (Map) o1;
+                for (Map.Entry<String, Object> e : map2.entrySet()) {
+                    final Object o2 = e.getValue();
+                    if (!(o2 instanceof Map)) {
+                        throw new ParsingException(p.getTokenLocation(),
+                            "Nested field mapping at " + index + "." + typeObjectEntry.getKey() + "." + e.getKey()
+                                + " is not found");
+                    }
+                    final Map<String, Object> map3 = (Map) o2;
+
+                    final String fullName = (String) map3.get(FieldMappingMetaData.FULL_NAME.getPreferredName());
+                    final XContentBuilder jsonBuilder = jsonBuilder();
+                    final Map<String, ?> values = (Map<String, ?>) map3.get(FieldMappingMetaData.MAPPING.getPreferredName());
+                    jsonBuilder.map(values);
```

Comment:
here you may be able to use copyCurrentStructure

Reply:
unfortunately - we already call `parser.map()` - therefore `copyCurrentStructure` is not feasible here

Reply:
makes sense

Reply:
So, why call `map` in the first place? It feels like maybe we should use the pull parser a bit more, build the map as we go.